module2/gift-certificate-system/service/src/main/java/com/epam/esm/service/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,135 @@
+    }
+
+    @Override
+    public void deleteCertificate(Long id) {
+        Optional<GiftCertificate> certificate = giftCertificateDao.find(id);
+        if (certificate.isPresent()) {
+            GiftCertificate giftCertificate = certificate.get();
+            giftCertificate.getTags().forEach(tag -> tagDao.removeTagAndCertificate(tag.getId(), id));
+            giftCertificateDao.delete(id);
+        } else {
+            throw new GiftCertificateNotFoundException(MessageFormat.format("Gift certificate with id: {0} not found", id));
+        }
+    }
+
+    @Override
+    public List<GiftCertificate> getCertificatesByTagName(String tagName) {
+        List<GiftCertificate> certificateList = giftCertificateDao.getCertificatesByTagName(tagName);
+        if (certificateList.isEmpty()) {
+            throw new GiftCertificateNotFoundException(MessageFormat.format("Gift certificate with tag name: {0} was not found", tagName));
+        } else {
+            return certificateList;
+        }
+    }
+
+    @Override
+    public List<GiftCertificate> getCertificatesByPartOfName(String partOfName) {
+        List<GiftCertificate> certificateList = giftCertificateDao.getCertificatesByPartOfName(partOfName);
+        if (certificateList.isEmpty()) {
+            throw new GiftCertificateNotFoundException(MessageFormat.format("Gift certificate containing the name: {0} was not found", partOfName));
+        } else {
+            return certificateList;
+        }
+    }
+
+    @Override
+    public List<GiftCertificate> getCertificatesByPartOfDescription(String partOfDescription) {
+        List<GiftCertificate> certificateList = giftCertificateDao.getCertificatesByPartOfDescription(partOfDescription);
+        if (certificateList.isEmpty()) {
+            throw new GiftCertificateNotFoundException(MessageFormat.format("Gift certificate containing the description: {0} was not found", partOfDescription));
+        } else {
+            return certificateList;
+        }
+    }
+
+    @Override
+    public void addTagToCertificate(Long certificateId, Long tagId) {
+        Optional<GiftCertificate> certificate = giftCertificateDao.find(certificateId);
+        Optional<Tag> tag = tagDao.find(tagId);
+        if (!certificate.isPresent()) {
+            throw new GiftCertificateNotFoundException(MessageFormat.format("Gift certificate with id: {0} was not found", certificateId));
+        } else if (!tag.isPresent()) {
+            throw new TagNotFoundException(MessageFormat.format("Tag with id: {0} was not found", tagId));
+        } else {
+            tagDao.addNewTagAndToCertificate(tagId, certificateId);
+        }
+    }
+
+    @Override
+    public void removeTagFromCertificate(Long certificateId, Long tagId) {
+        giftCertificateDao.removeTagFromCertificate(certificateId, tagId);
+    }
+
+    @Override
+    public List<GiftCertificate> sortCertificateByParameters(String sortParameter, String direction, List<GiftCertificate> certificates) {
```

Comment:
This method has code duplication. You can sort collection as first action and then check direction parameter.

Reply:
Fixed